src/heartratetask/HeartRateTask.cpp

```diff
@@ -110,10 +90,173 @@ void HeartRateTask::StartMeasurement() {
+    case States::ScreenOffAndMeasuring:
+    case States::ScreenOffAndWaiting:
+      state = States::ScreenOffAndStopped;
+      StopMeasurement();
+      break;
+    case States::ScreenOnAndStopped:
+    case States::ScreenOffAndStopped:
+      // shouldn't happen -> ignore
+      break;
+  }
+}
+
+void HeartRateTask::HandleBackgroundWaiting() {
+  if (!IsBackgroundMeasurementActivated()) {
+    return;
+  }
+
+  TickType_t ticksSinceWaitingStart = xTaskGetTickCount() - backgroundWaitingStart;
+  if (ticksSinceWaitingStart >= GetHeartRateBackgroundMeasurementIntervalInTicks()) {
+    state = States::ScreenOffAndMeasuring;
+    StartMeasurement();
+  }
+}
+
+void HeartRateTask::HandleSensorData(int* lastBpm) {
+  int8_t ambient = ppg.Preprocess(heartRateSensor.ReadHrs(), heartRateSensor.ReadAls());
+  int bpm = ppg.HeartRate();
+
+  // If ambient light detected or a reset requested (bpm < 0)
+  if (ambient > 0) {
+    // Reset all DAQ buffers
+    ppg.Reset(true);
+  } else if (bpm < 0) {
+    // Reset all DAQ buffers except HRS buffer
+    ppg.Reset(false);
+    // Set HR to zero and update
+    bpm = 0;
+  }
+
+  if (*lastBpm == 0 && bpm == 0) {
+    controller.Update(Controllers::HeartRateController::States::NotEnoughData, bpm);
+  }
+
+  if (bpm != 0) {
+    *lastBpm = bpm;
+    controller.Update(Controllers::HeartRateController::States::Running, bpm);
+    if (state == States::ScreenOnAndMeasuring || IsContinuosModeActivated()) {
+      return;
+    }
+    if (state == States::ScreenOffAndMeasuring) {
+      state = States::ScreenOffAndWaiting;
+      StartWaiting();
+    }
+  }
+  TickType_t ticksSinceMeasurementStart = xTaskGetTickCount() - measurementStart;
+  if (bpm == 0 && state == States::ScreenOffAndMeasuring && !IsContinuosModeActivated() &&
+      ticksSinceMeasurementStart >= DURATION_UNTIL_BACKGROUND_MEASUREMENT_IS_STOPPED) {
+    state = States::ScreenOffAndWaiting;
+    StartWaiting();
+  }
+}
+
+TickType_t HeartRateTask::GetHeartRateBackgroundMeasurementIntervalInTicks() {
+  int ms;
+  switch (settings.GetHeartRateBackgroundMeasurementInterval()) {
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::TenSeconds:
+      ms = 10 * 1000;
+      break;
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::ThirtySeconds:
+      ms = 30 * 1000;
+      break;
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::OneMinute:
+      ms = 60 * 1000;
+      break;
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::FiveMinutes:
+      ms = 5 * 60 * 1000;
+      break;
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::TenMinutes:
+      ms = 10 * 60 * 1000;
+      break;
+    case Pinetime::Controllers::Settings::HeartRateBackgroundMeasurementInterval::ThirtyMinutes:
+      ms = 30 * 60 * 1000;
+      break;
+    default:
+      ms = 0;
+      break;
+  }
+  return pdMS_TO_TICKS(ms);
```

Comment:
It's not a problem here but it's better to multiply by `configTICK_RATE_HZ` when you have an integer number of seconds to avoid overflows

Reply:
when looking for the definition of `pdMS_TO_TICKS` instead of that macro `/ configTICKRATE_HZ` was used before
Which version should I use? Since I'm usually not an developing for embedded devices, I'm not sure which one is the best one.

Reply:
If you have an integer (whole) number of seconds, you can simplify the calculation to `numberOfSeconds * configTICK_RATE_HZ`. If you have a number of milliseconds, then you have to do the calculation in `pdMS_TO_TICKS`